docs/man/noti.1
.\" Automatically generated by Pandoc 2.13
.\"
.TH "NOTI" "1" "2018/03/25" "noti 3.5.0" "Noti Manual"
.hy
.SH NAME
.PP
noti - monitor a process and trigger a notification
.SH SYNOPSIS
.PP
noti [flags] [utility [args\&...]]
.SH DESCRIPTION
.PP
Never sit and wait for some long-running process to finish.
Noti can alert you when it\[cq]s done.
You can receive messages on your computer or phone.
.SH OPTIONS
.TP
-t <string>, --title <string>
Set notification title.
Default is utility name.
.TP
-m <string>, --message <string>
Set notification message.
Default is \[lq]Done!\[rq].
Read from stdin with \[lq]-\[rq].
.TP
-e, --time
Show command execution time in message.
.TP
-b, --banner
Trigger a banner notification.
This is enabled by default.
To disable this service, set this flag to false.
This will be either \f[C]nsuser\f[R], \f[C]freedesktop\f[R], or
\f[C]notifyicon\f[R] notification, depending on the OS.
.TP
-s, --speech
Trigger a speech notification.
This will be either \f[C]say\f[R], \f[C]espeak\f[R], or
\f[C]speechsynthesizer\f[R] notification, depending on the OS.
.TP
-c, --bearychat
Trigger a BearyChat notification.
This requires \f[C]bearychat.incomingHookURI\f[R] to be set.
.TP
\[en]keybase
Trigger a Keybase notification.
This requires \f[C]keybase.conversation\f[R] to be set.
.TP
-p, --pushbullet
Trigger a Pushbullet notification.
This requires \f[C]pushbullet.accessToken\f[R] to be set.
.TP
-o, --pushover
Trigger a Pushover notification.
This requires \f[C]pushover.apiToken\f[R] and \f[C]pushover.userKey\f[R]
to be set.
.TP
-u, --pushsafer
Trigger a Pushsafer notification.
This requires \f[C]pushsafer.key\f[R] to be set.
.TP
-l, --simplepush
Trigger a Simplepush notification.
This requires \f[C]simplepush.key\f[R] to be set.
.TP
-k, --slack
Trigger a Slack notification.
This requires \f[C]slack.appurl\f[R] (for Slack apps) or
\f[C]slack.token\f[R] and \f[C]slack.channel\f[R] (for legacy tokens) to
be set.
.TP
-g, --telegram
Trigger a Telegram notification.
This requires \f[C]telegeram.token\f[R] and \f[C]telegram.chatId\f[R] to
be set.
.TP
\[en]twilio
Trigger a Twilio notification.
This requires \f[C]twilio.authToken\f[R], \f[C]twilio.accountSid\f[R],
\f[C]twilio.numberFrom\f[R] and \f[C]twilio.numberTo\f[R] to be set.
.TP
-w <pid>, --pwatch <pid>
Monitor a process by PID and trigger a notification when the pid
disappears.
.TP
-f, --file
Path to \f[C]noti.yaml\f[R] configuration file.
.TP
--verbose
Enable verbose mode.
.TP
-v, --version
Print \f[C]noti\f[R] version and exit.
.TP
-h, --help
Print \f[C]noti\f[R] help and exit.
.SH ENVIRONMENT
.IP \[bu] 2
\f[C]NOTI_DEFAULT\f[R]
.IP \[bu] 2
\f[C]NOTI_NSUSER_SOUNDNAME\f[R]
.IP \[bu] 2
\f[C]NOTI_NSUSER_SOUNDNAMEFAIL\f[R]
.IP \[bu] 2
\f[C]NOTI_SAY_VOICE\f[R]
.IP \[bu] 2
\f[C]NOTI_ESPEAK_VOICENAME\f[R]
.IP \[bu] 2
\f[C]NOTI_SPEECHSYNTHESIZER_VOICE\f[R]
.IP \[bu] 2
\f[C]NOTI_BEARYCHAT_INCOMINGHOOKURI\f[R]
.IP \[bu] 2
\f[C]NOTI_KEYBASE_CONVERSATION\f[R]
.IP \[bu] 2
\f[C]NOTI_KEYBASE_CHANNEL\f[R]
.IP \[bu] 2
\f[C]NOTI_KEYBASE_PUBLIC\f[R]
.IP \[bu] 2
\f[C]NOTI_KEYBASE_EXPLODINGLIFETIME\f[R]
.IP \[bu] 2
\f[C]NOTI_PUSHBULLET_ACCESSTOKEN\f[R]
.IP \[bu] 2
\f[C]NOTI_PUSHBULLET_DEVICEIDEN\f[R]
.IP \[bu] 2
\f[C]NOTI_PUSHOVER_APITOKEN\f[R]
.IP \[bu] 2
\f[C]NOTI_PUSHOVER_USERKEY\f[R]
.IP \[bu] 2
\f[C]NOTI_PUSHSAFER_KEY\f[R]
.IP \[bu] 2
\f[C]NOTI_SIMPLEPUSH_KEY\f[R]
.IP \[bu] 2
\f[C]NOTI_SIMPLEPUSH_EVENT\f[R]
.IP \[bu] 2
\f[C]NOTI_SLACK_TOKEN\f[R]
.IP \[bu] 2
\f[C]NOTI_SLACK_CHANNEL\f[R]
.IP \[bu] 2
\f[C]NOTI_SLACK_USERNAME\f[R]
.IP \[bu] 2
\f[C]NOTI_TWILIO_TO\f[R]
.IP \[bu] 2
\f[C]NOTI_TWILIO_FROM\f[R]
.IP \[bu] 2
\f[C]NOTI_TWILIO_ACCOUNTSID\f[R]
.IP \[bu] 2
\f[C]NOTI_TWILIO_AUTHTOKEN\f[R]
.IP \[bu] 2
\f[C]NOTI_TELEGRAM_CHATID\f[R]
.IP \[bu] 2
\f[C]NOTI_TELEGRAM_TOKEN\f[R]
.SH FILES
.PP
If not explicitly set with --file, then noti will check the following
paths, in the following order.
.IP \[bu] 2
\&./.noti.yaml
.IP \[bu] 2
$XDG_CONFIG_HOME/noti/noti.yaml
.SH EXAMPLES
.PP
Display a notification when \f[C]tar\f[R] finishes compressing files.
.IP
.nf
\f[C]
noti tar -cjf music.tar.bz2 Music/
\f[R]
.fi
.PP
Add noti after a command, in case you forgot at the beginning.
.IP
.nf
\f[C]
clang foo.c -Wall -lm -L/usr/X11R6/lib -lX11 -o bizz; noti
\f[R]
.fi
.PP
If you already started a command, but forgot to use \f[C]noti\f[R], then
you can do this to get notified when that process\[cq] PID disappears.
.IP
.nf
\f[C]
noti --pwatch $(pgrep docker-machine)
\f[R]
.fi
.PP
Receive your message from stdin with \f[C]-\f[R].
.IP
.nf
\f[C]
rsync -az --stats \[ti]/  server:/backups/homedir | noti -t \[dq]backup stats\[dq] -m -
\f[R]
.fi
.SH REPORTING BUGS
.PP
Report bugs on GitHub at https://github.com/variadico/noti/issues.
.SH SEE ALSO
.PP
noti.yaml(5)
.SH AUTHORS
Jaime Pi\[~n]a.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [2.7.0]

## Deprecated

* Single-dash long options. Any scripts using `-banner` or `-title` should be
  updated to use `--banner` or `--title` instead.

## [2.6.0]

## Added

* Support for Pushsafer.

## [2.5.0]

## Added

* Support for Simplepush.

## Fixed

* Formatting bug in help.

## [2.4.0]

## Added

* Support for BearyChat.

## [2.3.0]

## Added

* Banner support for Windows 10.

## Fixed

* Aliases in Bash and ZSH now work.

## [2.2.2]

## Fixed

* `noti` now compiles on Windows.

## [2.2.1]

## Added

* Add notification config docs to repo.

## Fixed

* Bug that caused noti to not work on macOS Sierra.

## [2.2.0]

## Added
* `-pwatch` flag to trigger notification after PID disappears.
* Check for updates during `-v` flag.

## Changed
* Install command changed to: `go get -u github.com/variadico/noti/cmd/noti`

## [2.1.1]

## Added
* Tests.

## Fixed
* Setting `-t` or `-m` will now take precedence over utility name.

## Changed

* Slackbot icon is now a rocket.
* If utility fails, noti will exit 1.
* Utility name default now includes subcommand too.
* Improved error handling for notifiers that shell out.

## [2.1.0]

## Added

* `NOTI_SOUND_FAIL` tells Noti which sound to play for banner notifications
when a utility fails on OS X.
* `NOTI_SOUND` and `NOTI_SOUND_FAIL` can be set to `_mute` for silent
notifications.
* HipChat notifications.
* Pushover notifications.
* Noti Wiki.

## Changed

* On OS X, banner notifications will play different sounds depending on the
utility's success or failure, instead of the same sound for both.
* `NOTI_SLACK_DEST` no longer defaults to "#random". It must be manually set.

## [2.0.0]

## Added

* Speech notifications.
* Slack notifications.
* Optionally set default notification type(s) through `NOTI_DEFAULT` env var.
* Multi-notification support.
* Other configuration through environment variables.
* Contributing document.
* Change log.

## Changed

* On OS X, the notification sound must now be set in the environment variable,
`NOTI_SOUND`.
* On OS X, instead of AppleScript, banner notifications are triggered with
Object-C, which shows (nicer) Terminal icon.

## Removed

* `-f` flag for OS X. This caused unexpected behavior for people who use iTerm2.
* OS X-specific flags and usage text from Linux and FreeBSD help.

[Unreleased]: https://github.com/variadico/noti/compare/v2.7.0...dev
[2.7.0]: https://github.com/variadico/noti/compare/v2.6.0...v2.7.0
[2.6.0]: https://github.com/variadico/noti/compare/v2.5.0...v2.6.0
[2.5.0]: https://github.com/variadico/noti/compare/v2.4.0...v2.5.0
[2.4.0]: https://github.com/variadico/noti/compare/v2.3.0...v2.4.0
[2.3.0]: https://github.com/variadico/noti/compare/v2.2.2...v2.3.0
[2.2.2]: https://github.com/variadico/noti/compare/v2.2.1...v2.2.2
[2.2.1]: https://github.com/variadico/noti/compare/v2.2.0...v2.2.1
[2.2.0]: https://github.com/variadico/noti/compare/v2.1.1...v2.2.0
[2.1.1]: https://github.com/variadico/noti/compare/v2.1.0...v2.1.1
[2.1.0]: https://github.com/variadico/noti/compare/v2.0.0...v2.1.0
[2.0.0]: https://github.com/variadico/noti/compare/v1.3.0...v2.0.0


cmd/noti/noti.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"time"

	"github.com/variadico/noti"
	"github.com/variadico/noti/banner"
	"github.com/variadico/noti/bearychat"
	"github.com/variadico/noti/hipchat"
	"github.com/variadico/noti/pushbullet"
	"github.com/variadico/noti/pushover"
	"github.com/variadico/noti/pushsafer"
	"github.com/variadico/noti/simplepush"
	"github.com/variadico/noti/slack"
	"github.com/variadico/noti/speech"
)

const (
	defaultEnv = "NOTI_DEFAULT"
	version    = "v2.7.0"
)

func main() {
	log.SetFlags(0)

	title := flag.String("t", "", "")
	message := flag.String("m", "", "")
	showVersion := flag.Bool("v", false, "")
	showHelp := flag.Bool("h", false, "")
	pid := flag.Int("w", -1, "")

	// Notifications
	bannerNoti := flag.Bool("b", true, "")
	hipChatNoti := flag.Bool("i", false, "")
	pushbulletNoti := flag.Bool("p", false, "")
	pushoverNoti := flag.Bool("o", false, "")
	pushsaferNoti := flag.Bool("u", false, "")
	simplepushNoti := flag.Bool("l", false, "")
	slackNoti := flag.Bool("k", false, "")
	speechNoti := flag.Bool("s", false, "")
	bearychatNoti := flag.Bool("c", false, "")

	flag.StringVar(title, "title", "", "")
	flag.StringVar(message, "message", "", "")
	flag.IntVar(pid, "pwatch", -1, "")
	flag.BoolVar(showVersion, "version", false, "")
	flag.BoolVar(showHelp, "help", false, "")

	// Notifications
	flag.BoolVar(bannerNoti, "banner", true, "")
	flag.BoolVar(hipChatNoti, "hipchat", false, "")
	flag.BoolVar(speechNoti, "speech", false, "")
	flag.BoolVar(pushbulletNoti, "pushbullet", false, "")
	flag.BoolVar(pushoverNoti, "pushover", false, "")
	flag.BoolVar(pushsaferNoti, "pushsafer", false, "")
	flag.BoolVar(simplepushNoti, "simplepush", false, "")
	flag.BoolVar(slackNoti, "slack", false, "")
	flag.BoolVar(bearychatNoti, "bearychat", false, "")

	flag.Parse()

	if flags := legacyFlags(os.Args); len(flags) > 0 {
		log.Println("WARNING: Found legacy-style flags:", flags)
		log.Println("Please pass long options with two dashes, e.g. --banner")
		log.Println("Legacy-style flags will be removed in the future")
		log.Println("See 'noti --help' for more information")
	}

	if *showVersion {
		fmt.Println("noti", version)
		checkForUpdates()
		return
	}
	if *showHelp {
		flag.Usage()
		return
	}

	env := noti.OSEnv{}
	setDefaultNotifications(flag.CommandLine, env)

	var n noti.Params
	if *pid == -1 {
		n = utilityNotification(flag.CommandLine)
	} else {
		n = pwatchNotification(flag.CommandLine, *pid)
	}

	n.Config = env

	notis := []struct {
		trigger bool
		api     string
		notify  func(noti.Params) error
	}{
		{*bannerNoti, "", banner.Notify},
		{*hipChatNoti, hipchat.API, hipchat.Notify},
		{*pushbulletNoti, pushbullet.API, pushbullet.Notify},
		{*pushoverNoti, pushover.API, pushover.Notify},
		{*pushsaferNoti, pushsafer.API, pushsafer.Notify},
		{*simplepushNoti, simplepush.API, simplepush.Notify},
		{*speechNoti, "", speech.Notify},
		{*slackNoti, slack.API, slack.Notify},
		{*bearychatNoti, "", bearychat.Notify},
	}

	for _, nt := range notis {
		if !nt.trigger {
			continue
		}

		n.API = nt.api

		if err := nt.notify(n); err != nil {
			log.Println(err)
		}
	}

	if n.Failure {
		os.Exit(1)
	}
}

func checkForUpdates() error {
	// Draft releases and prereleases are not returned by this endpoint.
	const releaseAPI = "https://api.github.com/repos/variadico/noti/releases/latest"
	webClient := &http.Client{Timeout: 30 * time.Second}

	resp, err := webClient.Get(releaseAPI)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	var r struct {
		HTMLURL string `json:"html_url"`
		TagName string `json:"tag_name"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&r); err != nil {
		return err
	}

	fmt.Println("latest:", r.TagName)
	if r.TagName != version {
		fmt.Println("download:", r.HTMLURL)
	}

	return nil
}

// setDefaultNotifications read the user's config and set their defaults on a
// FlagSet.
func setDefaultNotifications(fl *flag.FlagSet, env noti.EnvGetter) {
	defs := strings.TrimSpace(env.Get(defaultEnv))
	if defs == "" {
		return
	}

	has := strings.Contains
	fl.Set("banner", fmt.Sprintf("%t", has(defs, "banner")))
	fl.Set("hipchat", fmt.Sprintf("%t", has(defs, "hipchat")))
	fl.Set("pushbullet", fmt.Sprintf("%t", has(defs, "pushbullet")))
	fl.Set("pushover", fmt.Sprintf("%t", has(defs, "pushover")))
	fl.Set("pushsafer", fmt.Sprintf("%t", has(defs, "pushsafer")))
	fl.Set("simplepush", fmt.Sprintf("%t", has(defs, "simplepush")))
	fl.Set("speech", fmt.Sprintf("%t", has(defs, "speech")))
	fl.Set("slack", fmt.Sprintf("%t", has(defs, "slack")))
}

func pwatchNotification(fl *flag.FlagSet, pid int) noti.Params {
	err := pollPID(pid, 1*time.Second)

	return noti.Params{
		Title:   notiTitle(fl, "noti", err),
		Message: notiMessage(fl, err),
		Failure: (err != nil),
	}
}

func utilityNotification(fl *flag.FlagSet) noti.Params {
	util, err := runUtility(fl.Args())

	return noti.Params{
		Title:   notiTitle(fl, util, err),
		Message: notiMessage(fl, err),
		Failure: (err != nil),
	}
}

func notiTitle(fl *flag.FlagSet, util string, err error) string {
	t := flagValue(fl, "t", "title")
	if t == "" {
		t = util
	}

	if err != nil {
		t = fmt.Sprintf("%s failed", t)
	}

	return t
}

func notiMessage(fl *flag.FlagSet, err error) string {
	m := flagValue(fl, "m", "message")
	if m == "" {
		m = "Done!"
	}

	if err != nil {
		m = err.Error()
	}

	return m
}

func flagValue(fl *flag.FlagSet, short, long string) string {
	var v string

	if userSet(fl, short) {
		v = fl.Lookup(short).Value.String()
	} else if userSet(fl, long) {
		v = fl.Lookup(long).Value.String()
	}

	return v
}

func runUtility(args []string) (string, error) {
	name := utilityName(args)
	if len(args) < 1 {
		return name, nil
	}

	if _, err := exec.LookPath(args[0]); err != nil {
		exp, expErr := expandAlias(args[0])
		if expErr != nil {
			return "", err
		}

		args = append(exp, args[1:]...)
	}

	cmd := exec.Command(args[0], args[1:]...)
	cmd.Stdin = os.Stdin
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	return name, cmd.Run()
}

func utilityName(args []string) string {
	switch len(args) {
	case 0:
		return "noti"
	case 1:
		return args[0]
	}

	// If the next arg isn't a flag.
	if args[1][0] != '-' {
		// Append a subcommand to the utility name.
		return fmt.Sprintf("%s %s", args[0], args[1])
	}

	return args[0]
}

// userSet returns true if a user passed a target flag. Otherwise, it returns
// false. With zero values, we can't easily tell if a user omited a flag or if
// she actually did passed it with a zero value.
func userSet(fl *flag.FlagSet, target string) bool {
	var explicitlySet bool

	fl.Visit(func(f *flag.Flag) {
		if f.Name == target {
			explicitlySet = true
		}
	})

	return explicitlySet
}

func expandAlias(a string) ([]string, error) {
	shell := os.Getenv("SHELL")

	cmd := exec.Command(shell, "-l", "-i", "-c", "which "+a)
	e, err := cmd.Output()
	if err != nil {
		return nil, err
	}

	exp := strings.TrimSpace(string(e))
	trimLen := fmt.Sprintf("%s: aliased to ", a)
	exp = exp[len(trimLen):]

	return strings.Split(exp, " "), nil
}

func legacyFlags(args []string) []string {
	legacy := map[string]struct{}{
		"-title":      struct{}{},
		"-message":    struct{}{},
		"-pwatch":     struct{}{},
		"-banner":     struct{}{},
		"-speech":     struct{}{},
		"-hipchat":    struct{}{},
		"-pushbullet": struct{}{},
		"-pushover":   struct{}{},
		"-pushsafer":  struct{}{},
		"-simplepush": struct{}{},
		"-slack":      struct{}{},
		"-bearychat":  struct{}{},
		"-version":    struct{}{},
		"-help":       struct{}{},
	}

	var flags []string
	for _, arg := range args {
		if _, found := legacy[arg]; found {
			flags = append(flags, arg)
		}
	}

	return flags
}
